sbt client (--client) does not print error output on repeated commands, just quits with "[error]..."

## steps

- Setup minimal project with sbt `1.5.5`
- with some main method that fails, lets say at runtime (but also happens sometimes with compiletime error output)
- have the sbt server powered down
- run `sbt --client runMain <your Main class>` and observe error output
- run the same command again -- **and just get the message that there was an error, but no details**
- run `sbt --client shutdown`, and repeat: the first time -- when the server boots up -- the error message is there, the second time run, it is not.

## problem

Reproducible on my machine only on recurring runs of `sbt --client` as described above, the thin client **sometimes doesn't produce any error output on runtime / compiletime errors**. Sometimes it works (on a running sbt instance) and then stops working later.

Example output: -- first time, error gets printed, second time, it doesn't get printed.

```
should print
[error] (run-main-0) java.lang.AssertionError: assertion failed
[error]
[error] obj1 == obj2
[error] |    |  |
[error] |    |  IS(4,blub)
[error] |    false
[error] IS(3,hellooooo)
[error] java.lang.AssertionError: assertion failed
[error]
[error] obj1 == obj2
[error] |    |  |
[error] |    |  IS(4,blub)
[error] |    false
[error] IS(3,hellooooo)
[error]
[error]         at com.eed3si9n.expecty.ExpectyBase$ExpectyListener.expressionRecorded(Expecty.scala:41)
[error]         at com.eed3si9n.expecty.RecorderRuntime.recordExpression(RecorderRuntime.scala:40)
[error]         at bci.logic.Stuff$.StuffMain(Stuff.scala:21)
[error]         at bci.logic.StuffMain.main(Stuff.scala:7)
[error] Total time: 1 s, completed Oct 21, 2021 6:18:05 PM
snuc@snuc-NUC8i7BEH:~/sandbox/bct/bci-logic$ sbt --client runMain bci.logic.StuffMain
[info] entering *experimental* thin client - BEEP WHIRR
[info] terminate the server with `shutdown`
> runMain bci.logic.StuffMain
[error] Total time: 0 s, completed Oct 21, 2021 6:18:07 PM
snuc@snuc-NUC8i7BEH:~/sandbox/bct/bci-logic$
```

## expectation

it should just print errors every time.

## notes

Maybe the client pipe gets closed too early or not flushed?

--- a/main/src/main/scala/sbt/internal/server/NetworkChannel.scala
+++ b/main/src/main/scala/sbt/internal/server/NetworkChannel.scala
@@ -602,6 +602,7 @@ final class NetworkChannel(
       execId: String,
       err: JsonRpcResponseError
   ): Unit = {
+    forceFlush()
     val m = JsonRpcResponseMessage("2.0", execId, None, Option(err))
     val bytes = Serialization.serializeResponseMessage(m)
     publishBytes(bytes)
@@ -611,13 +612,17 @@ final class NetworkChannel(
   private[sbt] def jsonRpcNotify[A: JsonFormat](method: String, params: A): Unit = {
     val m =
       JsonRpcNotificationMessage("2.0", method, Option(Converter.toJson[A](params).get))
-    if (method != Serialization.systemOut) log.debug(s"jsonRpcNotify: $m")
+    if (method != Serialization.systemOut) {
+      forceFlush()
+      log.debug(s"jsonRpcNotify: $m")
+    }
     val bytes = Serialization.serializeNotificationMessage(m)
     publishBytes(bytes)
   }
 
   /** Notify to Language Server's client. */
   private[sbt] def jsonRpcRequest[A: JsonFormat](id: String, method: String, params: A): Unit = {
+    forceFlush()
     val m =
       JsonRpcRequestMessage("2.0", id, method, Option(Converter.toJson[A](params).get))
     log.debug(s"jsonRpcRequest: $m")
@@ -654,7 +659,13 @@ final class NetworkChannel(
 
   import scala.collection.JavaConverters._
   private[this] val outputBuffer = new LinkedBlockingQueue[Byte]
-  private[this] val flushFuture = new AtomicReference[java.util.concurrent.Future[_]]
+  private[this] val flushExecutor = Executors.newSingleThreadScheduledExecutor(
+    r => new Thread(r, s"$name-output-buffer-timer-thread")
+  )
+  private[this] def forceFlush() = {
+    Util.ignoreResult(flushExecutor.shutdownNow())
+    doFlush()
+  }
   private[this] def doFlush()() = {
     val list = new java.util.ArrayList[Byte]
     outputBuffer.synchronized(outputBuffer.drainTo(list))
@@ -673,13 +684,9 @@ final class NetworkChannel(
      * probably long enough to catch each burst but short enough to not introduce
      * noticeable latency.
      */
-    private[this] val executor =
-      Executors.newSingleThreadScheduledExecutor(
-        r => new Thread(r, s"$name-output-buffer-timer-thread")
-      )
+    private[this] val flushFuture = new AtomicReference[java.util.concurrent.Future[_]]
     override def close(): Unit = {
-      Util.ignoreResult(executor.shutdownNow())
-      doFlush()
+      forceFlush()
     }
     override def write(b: Int): Unit = outputBuffer.synchronized {
       outputBuffer.put(b.toByte)
@@ -689,7 +696,7 @@ final class NetworkChannel(
         case null =>
           try {
             flushFuture.set(
-              executor.schedule(
+              flushExecutor.schedule(
                 (() => {
                   flushFuture.set(null)
                   doFlush()

